src/source_util.cpp

```diff
@@ -74,15 +65,165 @@ bool sourceRangeContainsOnly(clang::SourceLocation beginLocation,
   return true;
 }
 
+/**
+ * Record for a single macro expansion step, including the location of that
+ * expansion and the macro information at that point, if any.
+ */
+struct ExpansionFrame {
+  const clang::MacroInfo *macroInfo;
+  clang::SourceLocation location;
+};
+
+/**
+ * Return a sequence of macro expansion records for the given location at the
+ * front of a range, from least expanded (the macro definition) to most
+ * expanded (the macro use). Stop if any partial macro expansions are
+ * encountered.
+ */
+std::vector<ExpansionFrame>
+expandBeginLocation(clang::SourceLocation begin,
```

Comment:
This is basically the old `isPartialMacro` split in two, and saving the macro information and location at each iteration in a stack.